src/ir/auth_logic/universe_relation_insertion.cc

```diff
@@ -0,0 +1,249 @@
+// universes for all the speakers given the type environment. Essentially,
+// it loops through the list of constant principals, and then all the
+// constants of all types, and then generates a SaysAssertion that
+// adds the constant to the universe for each speaker.
+std::vector<SaysAssertion> GetUniverseDefiningFacts(
+    const TypeEnvironment& type_env) {
+  std::vector<Principal> constant_principals;
+  for (const auto& typing : type_env.literal_type_map()) {
+    if (typing.second.kind() == datalog::ArgumentType::Kind::kPrincipal) {
+      constant_principals.push_back(Principal(typing.first));
+    }
+  }
+
+  std::vector<SaysAssertion> universe_defining_says_assertions;
+  for (const auto& prin : constant_principals) {
+    for (const auto& typing : type_env.literal_type_map()) {
+      universe_defining_says_assertions.push_back(SaysAssertion(
+          prin,  // speaker
+          {Assertion(Fact(
+              {}, MakeUniverseMembershipFact(typing.first, typing.second)))}));
+    }
+  }
+  return universe_defining_says_assertions;
+}
+
+// This visitor is used only to implement GetUniverseDeclarations
+class UniverseDeclarationGenerationVisitor
+    : public AuthLogicAstTraversingVisitor<
+          UniverseDeclarationGenerationVisitor,
+          absl::flat_hash_set<datalog::RelationDeclaration>> {
+ private:
+  using DeclSet = absl::flat_hash_set<datalog::RelationDeclaration>;
+  DeclSet GetDefaultValue() override { return {}; }
+
+  DeclSet CombineResult(DeclSet left, DeclSet right) override {
+    left.merge(std::move(right));
+    return left;
+  }
+
+  DeclSet Visit(
+      const datalog::RelationDeclaration& relation_declaration) override {
+    DeclSet universe_declarations;
+    for (const auto& argument : relation_declaration.arguments()) {
+      universe_declarations.insert(
+          TypeToUniverseDeclaration(argument.argument_type()));
+    }
+    return universe_declarations;
+  }
+};
+
+static absl::flat_hash_set<datalog::RelationDeclaration>
+GetUniverseDeclarations(const Program& prog) {
+  UniverseDeclarationGenerationVisitor gen;
+  absl::flat_hash_set<datalog::RelationDeclaration> ret = prog.Accept(gen);
+  // It is possible that universes need to be created
+  // for numbers and principals even if they are not
+  // referenced in relation declarations in the front-end
+  absl::flat_hash_set<datalog::RelationDeclaration> builtin_universes = {
+      TypeToUniverseDeclaration(datalog::ArgumentType(
+          datalog::ArgumentType::Kind::kNumber, "Number")),
+      TypeToUniverseDeclaration(datalog::ArgumentType(
+          datalog::ArgumentType::Kind::kPrincipal, "Principal")),
+  };
+  ret.merge(builtin_universes);
+  return ret;
+}
+
+//----------------------------------------------------------------------------
+// Grounding Condition Transformation
+//----------------------------------------------------------------------------
+// This class transforms SaysAssertions so that if
```

Comment:
Add an example transformation in the comment?

Reply:
fixed. included one to show overconservativeness